src/Purebred/Storage/Addressbook.hs

```diff
@@ -0,0 +1,87 @@
+-- This file is part of purebred
+-- Copyright (C) 2022 - Róman Joost
+--
+-- purebred is free software: you can redistribute it and/or modify
+-- it under the terms of the GNU Affero General Public License as published by
+-- the Free Software Foundation, either version 3 of the License, or
+-- (at your option) any later version.
+--
+-- This program is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Affero General Public License for more details.
+--
+-- You should have received a copy of the GNU Affero General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Purebred.Storage.Addressbook
+  ( -- * Synopsis
+    -- $synopsis
+
+    -- * API
+    queryAddresses
+  , runMuttAliasRetriever
+
+    -- * Muttalias Parser
+  , parseMuttAliasFile
+  )
+where
+
+import Data.Attoparsec.ByteString (Parser, sepBy, takeTill, parseOnly, string)
+import Data.Attoparsec.ByteString.Char8 (endOfLine, space, isSpace_w8, skipSpace)
+import qualified Data.ByteString as B
+import qualified Data.Text as T
+import qualified Data.Text.Internal.Search as T
+
+import Control.Lens
+import Control.Monad.Except (MonadError)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+
+import Data.IMF (address, Address)
+import Data.MIME (defaultCharsets)
+
+import Purebred.Types.String (decodeLenient)
+import Purebred.Types.Error
+import Purebred.Types.Addressbook
+-- $synopsis
+--
+-- This module provides an API to retrieve Address information for Purebred
+
+-- | query addresses matching given needle
+--
+queryAddresses :: (MonadError Error m, MonadIO m) => T.Text -> [AddressRetriever] -> m [Address]
+queryAddresses needle retrievers =
+  traverse (go needle) retrievers >>= pure . concat
+  where
+    go n r =
+        let config = view arConfig r
+        in liftIO (view arRetrieve r config n)
+
+-- | XXXX dependency on the constructor
+runMuttAliasRetriever :: (MonadIO m) => AddressRetrieverConfig -> T.Text -> m [Address]
```

Comment:
@frasertweedale I hate this, since it makes it hard to expand. I wonder whether I'm missing something in the type 🤔 Would you have a suggestion what to improve here?

Reply:
@romanofski what do you mean by "makes it hard to expand"?

Reply:
As in being extensible. If one of the goals was to provide an extensible API I can foresee problems with the data types if I for example would want to add support for an LDAP retriever. Since the constructor might be different and while pattern matching would work, it feels as if we used something more flexible. Yet I can't think of what it was.

Reply:
Thanks @romanofski, I understand.  So, I think my comment at https://github.com/purebred-mua/purebred/pull/499#discussion_r1045429227 addresses this concern.  The idea is that the address retriever UI is "just" a `Text -> m [Address]`, and you give it its configuration when constructing it.  Then all address book implementations have a simple and uniform query API.

Reply:
Yeah that is definitely much better. I totally forgot about the possibility of partials (doh!). I like this much better now.